lib/at_shell/at_shell.c

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdlib.h>
+
+#include <shell/shell.h>
+#include <nrf_modem_at.h>
+#include <modem/at_monitor.h>
+
+static struct shell *global_shell;
+static const char at_usage_str[] =
+	"Usage: at <subcommand>\n"
+	"\n"
+	"Subcommands:\n"
+	"  events_enable     Enable AT event handler which prints AT notifications\n"
+	"  events_disable    Disable AT event handler\n"
+	"\n"
+	"Any other subcommand is interpreted as AT command and sent to the modem.\n";
+
+AT_MONITOR(at_shell_monitor, ANY, at_notif_handler, PAUSED);
+
+static void at_notif_handler(const char *response)
+{
+	shell_print(global_shell, "AT event handler: \n%s", response);
+}
+
+static void at_print_error_info(const struct shell *shell, int error)
+{
+	switch (nrf_modem_at_err_type(error)) {
+	case NRF_MODEM_AT_ERROR:
+		shell_print(shell, "ERROR");
+		break;
+	case NRF_MODEM_AT_CME_ERROR:
+		shell_print(shell, "CME ERROR: %d", nrf_modem_at_err(error));
+		break;
+	case NRF_MODEM_AT_CMS_ERROR:
+		shell_print(shell, "CMS ERROR: %d", nrf_modem_at_err(error));
+		break;
+	default:
+		break;
+	}
+}
+
+int at_shell(const struct shell *shell, size_t argc, char **argv)
+{
+	int err;
+	static char response[CONFIG_AT_SHELL_CMD_RESPONSE_MAX_LEN + 1];
+
+	if (global_shell == NULL) {
+		global_shell = (struct shell *)shell;
+	}
+
+	if (argc < 2) {
+		shell_print(shell, "%s", argc, at_usage_str);
+		return 0;
+	}
+
+	char *command = argv[1];
+
+	if (!strcmp(command, "events_enable")) {
+		at_monitor_resume(at_shell_monitor);
+		shell_print(shell, "AT command event handler enabled");
+	} else if (!strcmp(command, "events_disable")) {
+		at_monitor_pause(at_shell_monitor);
+		shell_print(shell, "AT command event handler disabled");
+	} else {
+		err = nrf_modem_at_cmd(response, sizeof(response), command);
+		if (err < 0) {
+			shell_print(shell, "Sending AT command failed with error code %d", err);
+			return err;
+		} else if (err) {
+			at_print_error_info(shell, err);
+			return -EINVAL;
+		}
+
+		shell_print(shell, "%s", response);
```

Comment:
As you mentioned in your comments, some application may want to use their own printing functions and not `shell_print()` directly to put some application specific formatting into all prints.
Giving a function pointer to use for printing would be a solution but there might still be issues with shell pointer as we avoid passing that in all mosh_print's in modem shell.

Reply:
I'll see if I can come up with a suggestion, possibly in a follow-up PR